tools/jsonstats.cpp

```diff
@@ -31,19 +31,107 @@ struct stat_s {
   size_t false_count;
   size_t byte_count;
   size_t structural_indexes_count;
+  size_t key_count;
+  size_t key_maximum_length;
+  size_t maximum_depth;
+  size_t ascii_key_count;
+  size_t ascii_string_count;
+  size_t maximum_object_size;
+  size_t maximum_array_size;
   bool valid;
 };
 
 using stat_t = struct stat_s;
 
+bool is_ascii(const char *p, size_t length) {
+  for (size_t i = 0; i < length; i++) {
+    if (static_cast<unsigned char>(p[i]) >= 128) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void recurse(simdjson::document::iterator &pjh, stat_t &answer, size_t depth) {
```

Comment:
To do this without an iterator, you should be able to recurse with something like:

```c++
void recurse(simdjson::document::element element, stat_t &answer, size_t depth) {
  auto [array, array_error] = element.as_array();
  if (!array_error) {
    for (auto child : array) {
      recurse(child, answer, depth+1);
    }
  } else {
    auto [object, object_error] = element.as_object();
    if (!object_error) {
      for (auto [key, value] : object) {
        recurse(value, answer, depth+1);
      }
    }
  } else {
    // scalar, no recursion
  }
}
```

Reply:
It's highlighting the need for some kind of a `get_type()` you can switch on, though.